rfcs/45-stageless.md

```diff
@@ -0,0 +1,759 @@
+- Port the rest of the engine and examples to the new API.
+
+Given the massive scope, that sounds relatively straightforward!
+However, doing so will break approximately the entire engine, and tests will not pass again until everything has been ported.
+
+Let's explore some implementation details.
+
+### Scheduling
+
+Systems and their scheduling information are bundled and stored.
+The actual dependency graph construction is deferred (which means scheduling methods are order-independent).
+
+```rust
+enum Order {
+    Before,
+    After
+}
+
+struct Config {
+    sets: HashSet<BoxedSystemLabel>,
+    edges: Vec<(Order, BoxedSystemLabel)>,
+    conditions: Vec<BoxedRunCondition>,
+}
+
+struct ConfiguredSystem {
+    system: BoxedSystem,
+    config: Config,
+    name: Option<BoxedSystemLabel>,
+}
+
+struct ConfiguredSet {
+    set: BoxedSystemLabel,
+    config: Config,
+}
+
+trait IntoConfiguredSystem<Params> {
+    fn configure(self) -> ConfiguredSystem;
+    fn configure_with(self, config: Config) -> ConfiguredSystem;
+    /// Configures the system to run before the system or set given by `label`.
+    fn before<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystem;
+    /// Configures the system to run after the system or set given by `label`.
+    fn after<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystem;
+    /// Configures the system to run as part of `set`.
+    fn in_set(self, set: impl SystemLabel) -> ConfiguredSystem;
+    /// Configures the system to run only if `condition` returns `true`.
+    fn run_if<P>(self, condition: impl IntoRunCondition<P>) -> ConfiguredSystem;
+    /// Sets `name` as the unique name for this system.
+    fn named(self, name: impl SystemLabel) -> ConfiguredSystem;
+}
+
+trait IntoConfiguredSet {
+    fn configure(self) -> ConfiguredSet;
+    fn configure_with(self, config: Config) -> ConfiguredSet;
+    /// Configures the system set to run before the system or set given by `label`.
+    fn before<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSet;
+    /// Configures the system set to run after the system or set given by `label`.
+    fn after<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSet;
+    /// Configures the system set to run as part of `set`.
+    fn in_set(self, set: impl SystemLabel) -> ConfiguredSet;
+    /// Configures the system set to run only if `condition` returns `true`.
+    fn run_if<P>(self, condition: impl IntoRunCondition<P>) -> ConfiguredSet;
+}
+```
+
+### Scheduling errors
+
+There are several reasons an app may error due to improper scheduling. Most are related to graph solvability.
+
+1. Your dependencies contain a loop or cycle.
+2. Your hierarchy contains a loop or cycle.
+3. Your hierarchy has a transitive edge (something is both in some set and not at the same time.)
+4. You called `.in_set` with a system's label.
+5. (User-configurable) You have a dependency between two things that aren't in the same set.

```

Comment:
(5) is an error when you add a dependency between systems in 2 different sets, but the sets don't have an explicit order. We could force the 2 sets into the order of the systems, but it is probably better to explicitly order the sets instead to make sure the user isn't adding ordering between systems that they don't mean to.

It doesn't quite force a strict hierarchy like I said above.

Reply:
In other words, (5) says that a dependency should not cross a set boundary.

> It doesn't quite force a strict hierarchy like I said above.

It doesn't force a strict hierarchy since a system in a set can be a dependency of a system in a sub-set:

```
[hierarchy]

- Set1
  - SystemA
  - Set2
    - SystemB

[dependencies]

# Ok, according to (5): both systems are under Set1
SystemA -> SystemB
```

Reply:
Sorry, I didn't describe this very well.

I think it would be good practice to only create dependencies between things that belong to a common parent set. That was my intent with (5). 

Taking Nilirad's example, the "problem" with defining a relationship between `SystemA` and `SystemB` is that said graph edge would only appear in a visualizer when the Set1 graph is flattened.